conn.go

```diff
@@ -262,71 +261,95 @@ func (cc *Conn) flush(genID uint32) error {
 		return err
 	}
 
-	messages, err := conn.SendMessages(batch)
+	sentMsgs, err := conn.SendMessages(batch)
 	if err != nil {
 		return fmt.Errorf("SendMessages: %w", err)
 	}
 
 	var errs error
 
-	// Fetch replies. Each message with the Echo flag triggers a reply of the same
-	// type. Additionally, if the first message of the batch has the Echo flag, we
-	// get a reply of type NFT_MSG_NEWGEN, which we ignore.
-	replyIndex := 0
-	for replyIndex < len(cc.messages) && cc.messages[replyIndex].Header.Flags&netlink.Echo == 0 {
-		replyIndex++
-	}
-	replies, err := conn.Receive()
-	for err == nil && len(replies) != 0 {
-		reply := replies[0]
-		if reply.Header.Type == netlink.Error && reply.Header.Sequence == messages[1].Header.Sequence {
-			// The next message is the acknowledgement for the first message in the
-			// batch; stop looking for replies.
+	seqToMsgMap := cc.getSeqToMsgMap(sentMsgs)
+
+	for {
+		ready, err := cc.isReadReady(conn)
+		if err != nil {
+			return err
+		}
+
+		if !ready {
 			break
-		} else if replyIndex < len(cc.messages) {
-			msg := messages[replyIndex+1]
-			if msg.Header.Sequence == reply.Header.Sequence && msg.Header.Type == reply.Header.Type {
-				// The only messages which set the echo flag are rule create messages.
-				err := cc.messages[replyIndex].rule.handleCreateReply(reply)
-				if err != nil {
-					errs = errors.Join(errs, err)
-				}
-				replyIndex++
-				for replyIndex < len(cc.messages) && cc.messages[replyIndex].Header.Flags&netlink.Echo == 0 {
-					replyIndex++
-				}
-			}
 		}
-		replies = replies[1:]
-		if len(replies) == 0 {
-			replies, err = conn.Receive()
+
+		replies, err := conn.Receive()
+		if err != nil {
+			errs = errors.Join(errs, fmt.Errorf("receive: %w", err))
 		}
-	}
 
-	// Fetch the requested acknowledgement for each message we sent.
-	for i := range cc.messages {
-		if i != 0 {
-			_, err = conn.Receive()
+		if len(replies) == 0 && cc.TestDial != nil {
+			// When using a test dial function, we don't always get a reply for each
+			// sent message. Additionally, there is no buffer to poll for more data,
+			// so we stop here.
+			break
 		}
-		if err != nil {
-			if errors.Is(err, os.ErrPermission) || errors.Is(err, syscall.ENOBUFS) || errors.Is(err, syscall.ENOMEM) {
-				// Kernel will only send one error to user space.
-				return err
+
+		for _, reply := range replies {
+			if err = cc.handleEchoReply(seqToMsgMap, reply); err != nil {
```

Comment:
nit: use `err :=` to scope the `err` variable to just this conditional (otherwise, I always wonder if there is intentional shadowing going on)